### PR TITLE
Correction Accept headers

### LINKED ISSUE #30171: HttpClient still forcing spaces into Accept headers
HttpClient forcing spaces into Accept headers with parameters breaking non-conforming REST.
When setting httpClient accept header.  Strings are forced with a space between the ;
This as been reported here :
https://github.com/dotnet/corefx/issues/18449
https://github.com/dotnet/corefx/issues/22472

But still happens in version 2.2
```c#
 static void Main(string[] args)
        {
            var httpClient = new HttpClient();
            var request = new HttpRequestMessage(HttpMethod.Get, "http://google.com");
            request.Headers.Add("Accept", "application/json;version=1");
            var result = httpClient.SendAsync(request).Result;
            Console.ReadLine();
        }
```

If you check Fiddler, accept is : "application/json; version=1" and not "application/json;version=1" , there's a space after ; 

Unfortunately some API's don't work well with this space.  Is there any workaround?

WebRequest have the same behaviour as well. witch didn't happen in net  framework 4.5 
```c#
var webRequest = (HttpWebRequest)System.Net.WebRequest.Create($"http://google.com");
webRequest.Method = "POST";
webRequest.Accept = "application/json;version=1";
webRequest.GetResponse().GetResponseStream();
```

### PR DESCRIPTION
**Pull request by williammagno**
The space between Accept parameters causes errors and call failures, making Accept unrecognizable. This same problem does not occur in .NetFramework. Resolves #30171

### CODE REVIEW

**Comment by MihaZupan** [comment]
@WillMagno thank you for contributing!

I changed the target branch to master. Please remove the commit you picked up from the 31GA branch so only your change is included in this PR

File: src/installer/pkg/packaging/snaps/dotnet-runtime-3.1/snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      - libgssapi-krb5-2
+      - liblttng-ust0
+      - libstdc++6
+      - zlib1g
+      - libgcc1
+      - lldb
+      - libunwind8
+      - libtinfo5
+      - libdb5.3
+      - libc6
```
Comment (stephentoub) [inline src/installer/pkg/packaging/snaps/dotnet-runtime-3.1/snap/snapcraft.yaml]: Why is this PR adding this and changing the below yaml file?

Comment (MihaZupan) [reply src/installer/pkg/packaging/snaps/dotnet-runtime-3.1/snap/snapcraft.yaml]: The PR originally targeted the 31GA-snap branch. That commit was picked up when I changed it to master.

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs [hunk 2]
```diff
@@ -180,7 +180,6 @@ private void AddToStringBuilder(StringBuilder sb)
                 if (leadingSeparator || (destination.Length > 0))
                 {
                     destination.Append(separator);
-                    destination.Append(' ');
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: This is going to affect much more than just Accept.  That's ok?

Also, I'm surprised no tests need to change; we don't have any tests validating the output here?

@davidsh, you had several comments on the original issue stating the current behavior was desirable; did we reverse our decision on that?

Comment (WillMagno) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: @stephentoub , I tested and researched some http request scenarios and there was no problem. Can you identify any scenarios that this change may cause problems?
If you deem it necessary to maintain the current behavior, I create an argument to use the space between the headers or not, so there will be no impact on the current function.

Comment (davidsh) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: I'm still seeing this file in the PR:
 src/installer/pkg/packaging/snaps/dotnet-sdk/snap/snapcraft.yaml

Comment (davidsh) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/NameValueHeaderValue.cs]: >This is going to affect much more than just Accept. That's ok?

I don't think this change is the correct one to make. It is overly broad and will affect a lot of headers.

**Comment by davidsh** [comment]
This PR is attempting to fix the issue reported in https://github.com/dotnet/corefx/issues/39260.

But .NET Core is following the RFC. And while we are generally flexible in what we receive on the wire, we try to be strict about what we send on the wire. Specifically we follow all 'MUST' requirements from RFC's. In in the majority of cases, we follow all 'SHOULD' requirements unless there is a compelling reason why we can't.

In this case, we are following the spec 'SHOULD' statement below:

```
The OWS rule is used where zero or more linear whitespace octets
might appear.  For protocol elements where optional whitespace is
preferred to improve readability, a sender SHOULD generate the
optional whitespace as a single SP
```

One could make an argument that we don't need (or should not) follow this part of the spec since 'readability' is not really an issue here.

But I worry that this PR is too broad and will potentially break other things.

**Comment by WillMagno** [comment]
@davidsh What if I create an argument indicating the use of space or not?

**Comment by davidsh** [comment]
>

The method being modified is an internal method 'ToString' that is probably used to serialize each headervalue. So, yes, the argument parameters could be modified since it is an internal method. But that change would have to rippled throughout the source code. And then the question is when should the call sites use a different behavior?

```
internal static void ToString(ObjectCollection<NameValueHeaderValue> values, char separator, bool leadingSeparator, StringBuilder destination)
```

Looking at the source code, it probably would have made sense if the method originally took `string separator` instead of `char separator`. That would give it more flexibility potentially.  

Write now it has the hard-code behavior of using the separator character and always adding a single 'space' as well.

But I still don't understand how changing this signature will help scope down the changes in general.

It was mentioned that .NET Framework behaves differently and doesn't emit a space character after the ';' separator. Is this true for all headers in HTTP or just some?

**Comment by ClimberBear** [comment]
Hi,

IMHO, we are misunderstanding the issue.

The problem is not to add the string separator, or the char separator.

The correct behavior is to keep the string UNCHANGED.

If the programmer write
`request.Accept="Accept:application/xml;notice=tree" `
the run time has to send as is (without spaces between the semicolon and notice.
Otherwise, if the programmer use an space, has to send a space, and so on.

I imagine (I have not checked the code) that the routine, when the Access property is set, is parsing the string, splitting by comma (,) and semi-colon (;) and adding it to a collection of elements. After, it uses this collection to rebuild the header.

This is not necessary, and it's creating the problem. Noone is going to add partial attributes in an Accept header. Maybe it can split by comma (if somebody calls Accept several times), but never by semi-colon.

Thanks.

**Comment by davidsh** [comment]
>The correct behavior is to keep the string UNCHANGED.

When developers use the strongly typed properties such as `.Accept`:

```c#
request.Headers.Add("Accept", "application/json;version=1");
```

.NET Core will generally normalize the input as you describe. And in this case, it will add the spaces after the ';' separator.

However, if developers instead use these APIs:

```c#
request.Headers.TryAddWithoutValidation("Accept", "application/json;version=1");
```

.NET Core is supposed to leave the string as-is and not validate it. It then should transmit it exactly as-is on the wire. If 'TryAddWithoutValidation' is still normalizing the input, then that would be a bug we should fix.

**Comment by stephentoub** [comment]
> .NET Core is supposed to leave the string as-is and not validate it. It then should transmit it exactly as-is on the wire. If 'TryAddWithoutValidation' is still normalizing the input, then that would be a bug we should fix.

That was fixed in https://github.com/dotnet/corefx/pull/41640 for 5.0.

**Comment by WillMagno** [comment]
In my view the core include space is still incorrect. If the developer wants to validate the header it will still be incorrect.

**Comment by WillMagno** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages

See the body of the requests, the response has spaces the request no.

**Comment by ClimberBear** [comment]
> > .NET Core is supposed to leave the string as-is and not validate it. It then should transmit it exactly as-is on the wire. If 'TryAddWithoutValidation' is still normalizing the input, then that would be a bug we should fix.
> 
> That was fixed in [dotnet/corefx#41640](https://github.com/dotnet/corefx/pull/41640) for 5.0.

Hi
We are using HttpWebRequest class, where:
1) We cannot use directly Headers.Add("Accept") directly because it has a specific property Accept
2) And has no TryAddWithoutValidation method

**Comment by WillMagno** [comment]
https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Accept

In the syntax, the ';' has no space. The ',' have.

I finished all the adjustments.

**Comment by WillMagno** [comment]
Hello @davidsh, I changed the code to treat semicolon as separator without adding the space. I also adjusted the tests to fit the new treatment.

**Comment by stephentoub** [comment]
@dotnet/ncl, what should happen with this PR?

**Comment by halter73** [comment]
Given that HttpWebRequest doesn't support TryAddWithoutValidation for Accept headers, and this matches the .NET Framework behavior, I would take this. Also, in my testing, neither Chrome nor Firefox adds a space after semicolons in their Accept(-Language) headers FWIW.

**Comment by wfurt** [comment]
Same for Safari. The scary part IMHO is that this change is not limited to `Accept ` and it can possibly impact many other headers.

**Comment by WillMagno** [comment]
If you want, I change it to affect only the ACCEPT header.

**Comment by scalablecory** [comment]
This form of `Accept` is already supported in 5.0 for `HttpWebRequest`, as it already uses `TryAddWithoutValidation` and the `TryAddWithoutValidation` behavior was changed per https://github.com/dotnet/runtime/pull/792#issuecomment-565468418.

That said, I'm not against this PR if we have it only affect the `Accept` header: Although RFC 7231 amended HTTP/1.1 to allow spaces, the original RFC 2616 did not support a space. Given that the space is optional in RFC 7231, to provide maximum compatibility with those interpreting the original spec I think it is better to ignore the OWS "SHOULD" and not send a space.

**Comment by scalablecory** [comment]
Actually I was mistaken, RFC 2616 has an "implied *LWS" rule in Section 2.1:

```
   implied *LWS
      The grammar described by this specification is word-based. Except
      where noted otherwise, linear white space (LWS) can be included
      between any two adjacent words (token or quoted-string), and
      between adjacent words and separators, without changing the
      interpretation of a field. At least one delimiter (LWS and/or
      separators) MUST exist between any two tokens (for the definition
      of "token" below), since they would otherwise be interpreted as a
      single token.
```

The OWS change in RFC 7231 only made this more explicit. So, we are in compliance and backward-compatible with the older RFC.

Given this, I am against taking this PR. It is not appropriate to change implementation to work around a few buggy web services, and we have a solution in `TryAddWithoutValidation` for those unable to fix their web service.

**Comment by ClimberBear** [comment]
Please, I'm not able to follow the thread. Is it going to be solved in Net Core 3.x (with or without TryAddWithoutValidation). Currently still not work!

**Comment by scalablecory** [comment]
Triage: we'd like to know when this change was introduced (.NET Core 1.0 vs SocketsHttpHandler). There may also be a perf benefit to not normalizing valid headers.

**Comment by MihaZupan** [comment]
That behaviour has been introduced to corefx in dotnet/corefx#1586. Will take a look into tfs history to see if I can narrow it down more.

**Comment by stephentoub** [comment]
> The same problem does not occur with .Net Framework

I just ran this in a .NET Framework console app:
```C#
using System;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;

class Program
{
    static void Main()
    {
        var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
        listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
        listener.Listen(1);

        var httpClient = new HttpClient();
        var request = new HttpRequestMessage(HttpMethod.Get, $"http://localhost:{((IPEndPoint)listener.LocalEndPoint).Port}");
        request.Headers.Add("Accept", "application/json;version=1");
        httpClient.SendAsync(request);

        var server = listener.Accept();
        var reader = new StreamReader(new NetworkStream(server));
        while (true) Console.WriteLine(reader.ReadLine());
    }
}
```
and I get back this:
```
GET / HTTP/1.1
Accept: application/json; version=1
Host: localhost:64592
Connection: Keep-Alive
```
so it certainly seems like .NET Framework has the same behavior.  You see otherwise?

**Comment by stephentoub** [comment]
@scalablecory, @karelz, it'd be good to circle back to this the next time you go through networking issues, and make a call on the next step.

My above test shows this isn't specific to .NET Core.  @WillMagno, you suggested it is, saying "This same problem does not occur in .NetFramework"... can you share a repro for the issue that happens on .NET Core but not on .NET Framework? Thanks!

**Comment by stephentoub** [comment]
@karelz, @scalablecory, it's been a month... has this been forgotten?  Thanks.

**Comment by karelz** [comment]
Yes, this has slipped through the cracks :(
@WillMagno can you please share details about differences of .NET Framework and .NET Core as asked above?

**Comment by WillMagno** [comment]
@karelz this was based on the WebRequest of .NetFramework 4.5.

https://github.com/dotnet/runtime/issues/30171

**Comment by lukehook** [comment]
Is there any update on this one? I've come across multiple REST API's that also don't support the addition of whitespaces. Am I right in the understanding that the TryWithoutValidation method should leave this unaltered?

**Comment by scalablecory** [comment]
> I've come across multiple REST API's that also don't support the addition of whitespaces.

Can you give some examples? This sort of data helps shape how we treat this.

> Am I right in the understanding that the TryWithoutValidation method should leave this unaltered?

It will in .NET 5. This is our current recommendation.

**Comment by scalablecory** [comment]
I'm going to close this PR, with a few notes and summary:

Primarily, this PR as implemented has an overly broad effect: it affects all headers with name-value parameters, not just the Accept header in question. This is not the change we would accept to address the issue in question.

Further, we determined that we are in compliance with HTTP and match the behavior of Framework. We have some concern that fixing one non-compliant app might break another non-compliant app. To help us understand the scope, I'd like to see some specific examples of web apps which break due to this. Please continue discussion in https://github.com/dotnet/runtime/issues/30171.

Our current recommendation for this scenario is to use `TryAddWithoutValidation`, which in .NET 5 (per https://github.com/dotnet/corefx/pull/41640) will not perform any validation or normalization. `HttpWebRequest` uses this behind the scenes, so this scenario should work there too. Some .NET 5 preview releases are out now, so if you are having this problem it's a great time to give that a try to ensure your scenario works in the coming release.

**Pull request closed without merging**